common/task/weightedRoundRobinTaskScheduler.go

```diff
@@ -0,0 +1,197 @@
+func NewWeightedRoundRobinTaskScheduler(
+	logger log.Logger,
+	metricsScope metrics.Scope,
+	options *WeightedRoundRobinTaskSchedulerOptions,
+) (Scheduler, error) {
+	if len(options.Weights) == 0 {
+		return nil, errors.New("weight is not specified in the scheduler option")
+	}
+
+	numPriorities := len(options.Weights)
+	scheduler := &weightedRoundRobinTaskSchedulerImpl{
+		status:       common.DaemonStatusInitialized,
+		numQueues:    numPriorities,
+		taskChs:      make([]chan PriorityTask, numPriorities),
+		shutdownCh:   make(chan struct{}),
+		notifyCh:     make(chan struct{}, 1),
+		logger:       logger,
+		metricsScope: metricsScope,
+		options:      options,
+		processor: NewParallelTaskProcessor(
+			logger,
+			metricsScope,
+			&ParallelTaskProcessorOptions{
+				QueueSize:   wRRTaskProcessorQueueSize,
+				WorkerCount: options.WorkerCount,
+				RetryPolicy: options.RetryPolicy,
+			},
+		),
+	}
+
+	for i := 0; i < numPriorities; i++ {
+		scheduler.taskChs[i] = make(chan PriorityTask, options.QueueSize)
+	}
+
+	return scheduler, nil
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&w.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	w.processor.Start()
+
+	w.dispatcherWG.Add(1)
+	go w.dispatcher()
+
+	w.logger.Info("Weighted round robin task scheduler started.")
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&w.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(w.shutdownCh)
+
+	w.processor.Stop()
+
+	if success := common.AwaitWaitGroup(&w.dispatcherWG, time.Minute); !success {
+		w.logger.Warn("Weighted round robin task scheduler timedout on shutdown.")
+	}
+
+	w.logger.Info("Weighted round robin task scheduler shutdown.")
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) Submit(task PriorityTask) error {
+	w.metricsScope.IncCounter(metrics.PriorityTaskSubmitRequest)
+	sw := w.metricsScope.StartTimer(metrics.PriorityTaskSubmitLatency)
+	defer sw.Stop()
+
+	priority := task.Priority()
+	if priority >= w.numQueues {
+		return errors.New("task priority exceeds limit")
+	}
+	select {
+	case w.taskChs[priority] <- task:
+		select {
+		case w.notifyCh <- struct{}{}:
+			// sent a notification to the dispatcher
+		default:
+			// do not block if there's already a notification
+		}
+		return nil
+	case <-w.shutdownCh:
+		return ErrTaskSchedulerClosed
+	}
+}
+
+func (w *weightedRoundRobinTaskSchedulerImpl) dispatcher() {
+	defer w.dispatcherWG.Done()
+
+	outstandingTasks := false
+
+	for {
+		if !outstandingTasks {
+			// if no task is dispatched in the last round,
+			// wait for a notification
+			select {
+			case <-w.notifyCh:
+				// block until there's a new task
+			case <-w.shutdownCh:
+				return
+			}
+		}
+
+		outstandingTasks = false
+		for priority := 0; priority != w.numQueues; priority++ {
+			for i := 0; i != w.options.Weights[priority]; i++ {
```

Comment:
It would be dead loop if someone mistakenly had negative Weight. Maybe safer to use `i < w.options.Weights[priority]`

Reply:
Updated.